File: backend/tellmenow/api/publish.py
import uuid
from datetime import datetime, timezone
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException

from tellmenow.api.auth import User, get_current_user, get_optional_user
from tellmenow.api.schemas import PublishRequest, PublishResponse, PublishedPageResponse
from tellmenow.db import get_db
from tellmenow.jobs.manager import job_manager

router = APIRouter(prefix="/api", tags=["publish"])


@router.post("/publish", response_model=PublishResponse)
async def publish_page(
    request: PublishRequest,
    user: Annotated[User, Depends(get_current_user)],
):
    job = job_manager.get_job(request.job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")

    if job.user_id and job.user_id != user.id:
        raise HTTPException(status_code=404, detail="Job not found")

    if not job.html_report:
        raise HTTPException(status_code=400, detail="Job has no HTML report to publish")

    published_id = uuid.uuid4().hex[:10]
    now = datetime.now(timezone.utc).isoformat()

    db = get_db()
    await db.execute(
        """INSERT INTO published_pages (id, job_id, user_id, title, html, skill_id, query, created_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?)""",
        (published_id, job.id, user.id, job.report_title or job.query[:80], job.html_report, job.skill_id, job.query, now),
    )
    await db.commit()

    job.published_id = published_id

    return PublishResponse(
        published_id=published_id,
        url=f"/p/{published_id}",
    )


@router.get("/p/{page_id}", response_model=PublishedPageResponse)
async def get_published_page(page_id: str):
    """Get a published page. No auth required — published pages are public."""
    db = get_db()
    cursor = await db.execute(
        "SELECT id, title, html, skill_id, query, created_at FROM published_pages WHERE id = ?",
        (page_id,),
    )
    row = await cursor.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Page not found")

    return PublishedPageResponse(
        id=row[0],
        title=row[1],
        html=row[2],
        skill_id=row[3],
        query=row[4],
        created_at=row[5],
    )

File: worker/src/routes/publish.ts
import { Hono } from "hono";
import { Env } from "../types";
import { getJob, insertPublishedPage, getPublishedPage } from "../db/queries";

const app = new Hono<{ Bindings: Env }>();

app.post("/publish", async (c) => {
  const body = await c.req.json<{ job_id: string }>();
  if (!body.job_id) {
    return c.json({ detail: "job_id is required" }, 400);
  }

  const job = await getJob(c.env.DB, body.job_id);
  if (!job) {
    return c.json({ detail: "Job not found" }, 404);
  }
  if (!job.html_report) {
    return c.json({ detail: "Job has no HTML report to publish" }, 400);
  }

  const publishedId = crypto.randomUUID().replace(/-/g, "").slice(0, 10);
  const now = new Date().toISOString();

  await insertPublishedPage(c.env.DB, {
    id: publishedId,
    job_id: job.id,
    user_id: null,
    title: job.report_title || job.query.slice(0, 80),
    html: job.html_report,
    skill_id: job.skill_id,
    query: job.query,
    created_at: now,
  });

  return c.json({
    published_id: publishedId,
    url: `/p/${publishedId}`,
  });
});

app.get("/p/:id", async (c) => {
  const page = await getPublishedPage(c.env.DB, c.req.param("id"));
  if (!page) {
    return c.json({ detail: "Page not found" }, 404);
  }

  return c.json({
    id: page.id,
    title: page.title,
    html: page.html,
    skill_id: page.skill_id,
    query: page.query,
    created_at: page.created_at,
  });
});

export default app;
